File: internal/cli/ipc.go
package cli

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"sync"
	"time"

	"github.com/google/uuid"
)

// JSON-RPC 2.0 protocol types for inter-process communication.
// Protocol: Unix domain socket, newline-delimited JSON (D-030).

// IPCRequest is a JSON-RPC 2.0 request message.
type IPCRequest struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      string          `json:"id"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params,omitempty"`
}

// IPCResponse is a JSON-RPC 2.0 response message.
type IPCResponse struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      string          `json:"id"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *IPCError       `json:"error,omitempty"`
}

// IPCError represents a JSON-RPC 2.0 error object.
type IPCError struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Data    json.RawMessage `json:"data,omitempty"`
}

// Error implements the error interface for IPCError.
func (e *IPCError) Error() string {
	return fmt.Sprintf("ipc error %d: %s", e.Code, e.Message)
}

// NewIPCRequest constructs a new IPCRequest with an auto-generated UUID v4 id.
func NewIPCRequest(method string, params any) (*IPCRequest, error) {
	var raw json.RawMessage
	if params != nil {
		b, err := json.Marshal(params)
		if err != nil {
			return nil, fmt.Errorf("new ipc request marshal params: %w", err)
		}
		raw = b
	}
	return &IPCRequest{
		JSONRPC: "2.0",
		ID:      uuid.New().String(),
		Method:  method,
		Params:  raw,
	}, nil
}

// NewIPCResponse constructs a successful IPCResponse with the given id and result.
func NewIPCResponse(id string, result any) (*IPCResponse, error) {
	raw, err := json.Marshal(result)
	if err != nil {
		return nil, fmt.Errorf("new ipc response marshal result: %w", err)
	}
	return &IPCResponse{
		JSONRPC: "2.0",
		ID:      id,
		Result:  raw,
	}, nil
}

// NewIPCErrorResponse constructs an IPCResponse carrying an error.
func NewIPCErrorResponse(id string, code int, message string) *IPCResponse {
	return &IPCResponse{
		JSONRPC: "2.0",
		ID:      id,
		Error: &IPCError{
			Code:    code,
			Message: message,
		},
	}
}

// IPCHandlerFunc is the signature for a method handler on the IPC server.
type IPCHandlerFunc func(ctx context.Context, req *IPCRequest) (*IPCResponse, error)

// IPCServer is a Unix domain socket server that dispatches JSON-RPC 2.0
// requests to registered method handlers.
//
// Handlers must be registered via Register before Start is called.
// At runtime the handlers map is read-only and requires no locking.
type IPCServer struct {
	sockPath string
	handlers map[string]IPCHandlerFunc
	listener net.Listener
	ctx      context.Context
	cancel   context.CancelFunc
	wg       sync.WaitGroup
}

// NewIPCServer creates a new IPCServer bound to the given Unix socket path.
func NewIPCServer(sockPath string) *IPCServer {
	return &IPCServer{
		sockPath: sockPath,
		handlers: make(map[string]IPCHandlerFunc),
	}
}

// Register binds a handler to a method name. Must be called before Start.
func (s *IPCServer) Register(method string, handler IPCHandlerFunc) {
	s.handlers[method] = handler
}

// Start begins accepting connections on the Unix socket. It is non-blocking:
// the listener loop runs in a separate goroutine.
//
// Before listening it removes any stale socket at the path. After the
// listener is created the socket file is chmod'd to 0600.
func (s *IPCServer) Start(ctx context.Context) error {
	// Remove stale socket.
	if err := os.Remove(s.sockPath); err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("ipc server remove stale socket: %w", err)
	}

	ln, err := net.Listen("unix", s.sockPath)
	if err != nil {
		return fmt.Errorf("ipc server listen: %w", err)
	}

	if err := os.Chmod(s.sockPath, 0600); err != nil {
		ln.Close()
		return fmt.Errorf("ipc server chmod socket: %w", err)
	}

	s.listener = ln
	s.ctx, s.cancel = context.WithCancel(ctx)

	s.wg.Add(1)
	go s.acceptLoop()

	return nil
}

// acceptLoop accepts connections until the listener is closed.
func (s *IPCServer) acceptLoop() {
	defer s.wg.Done()
	for {
		conn, err := s.listener.Accept()
		if err != nil {
			// Listener closed — check context and exit.
			select {
			case <-s.ctx.Done():
				return
			default:
				// Backoff to avoid CPU spin on transient accept errors.
				time.Sleep(100 * time.Millisecond)
				continue
			}
		}
		s.wg.Add(1)
		go s.handleConn(conn)
	}
}

// handleConn reads newline-delimited JSON requests from a single connection,
// dispatches them to handlers, and writes back JSON responses.
func (s *IPCServer) handleConn(conn net.Conn) {
	defer s.wg.Done()
	defer conn.Close()

	scanner := bufio.NewScanner(conn)
	scanner.Buffer(make([]byte, 64*1024), 1024*1024)

	for scanner.Scan() {
		line := scanner.Bytes()
		resp := s.dispatch(line)
		b, err := json.Marshal(resp)
		if err != nil {
			// Should not happen with well-formed IPCResponse.
			continue
		}
		if _, err := conn.Write(append(b, '\n')); err != nil {
			return
		}
	}
	_ = scanner.Err()
}

// dispatch parses a single JSON-RPC request line and invokes the handler.
func (s *IPCServer) dispatch(line []byte) *IPCResponse {
	var req IPCRequest
	if err := json.Unmarshal(line, &req); err != nil {
		return NewIPCErrorResponse("", -32700, "Parse error")
	}
	if req.JSONRPC != "2.0" {
		return NewIPCErrorResponse(req.ID, -32600, "Invalid Request")
	}

	handler, ok := s.handlers[req.Method]
	if !ok {
		return NewIPCErrorResponse(req.ID, -32601, "Method not found")
	}

	resp, err := handler(s.ctx, &req)
	if err != nil {
		return NewIPCErrorResponse(req.ID, -32000, err.Error())
	}
	return resp
}

// Stop closes the listener and waits for all in-flight connections to finish.
func (s *IPCServer) Stop() error {
	if s.cancel != nil {
		s.cancel()
	}
	var err error
	if s.listener != nil {
		err = s.listener.Close()
	}
	s.wg.Wait()
	return err
}

// IPCClient is a client for the Unix socket IPC server. Each Call opens a
// fresh connection, sends one request, reads one response, and closes.
type IPCClient struct {
	sockPath string
	timeout  time.Duration
}

// NewIPCClient creates a new IPCClient for the given socket path with a
// connection timeout.
func NewIPCClient(sockPath string, timeout time.Duration) *IPCClient {
	return &IPCClient{
		sockPath: sockPath,
		timeout:  timeout,
	}
}

// Call sends a single JSON-RPC request to the IPC server and returns the
// response. It opens a new connection per call.
func (c *IPCClient) Call(ctx context.Context, method string, params any) (*IPCResponse, error) {
	conn, err := net.DialTimeout("unix", c.sockPath, c.timeout)
	if err != nil {
		return nil, fmt.Errorf("ipc client dial: %w", err)
	}
	defer conn.Close()

	// Set read deadline so we don't block indefinitely if the server never responds.
	if err := conn.SetReadDeadline(time.Now().Add(c.timeout)); err != nil {
		return nil, fmt.Errorf("ipc client set read deadline: %w", err)
	}

	req, err := NewIPCRequest(method, params)
	if err != nil {
		return nil, fmt.Errorf("ipc client build request: %w", err)
	}

	b, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("ipc client marshal request: %w", err)
	}
	b = append(b, '\n')

	if _, err := conn.Write(b); err != nil {
		return nil, fmt.Errorf("ipc client write request: %w", err)
	}

	scanner := bufio.NewScanner(conn)
	scanner.Buffer(make([]byte, 64*1024), 1024*1024)
	if !scanner.Scan() {
		if err := scanner.Err(); err != nil {
			return nil, fmt.Errorf("ipc client read response: %w", err)
		}
		return nil, fmt.Errorf("ipc client read response: no response received")
	}

	var resp IPCResponse
	if err := json.Unmarshal(scanner.Bytes(), &resp); err != nil {
		return nil, fmt.Errorf("ipc client unmarshal response: %w", err)
	}
	return &resp, nil
}

File: demo/web/packages/xyncra-client-cli/src/ipc.ts
/**
 * IPC (Inter-Process Communication) server and client over Unix domain sockets.
 *
 * Protocol: JSON-RPC 2.0, newline-delimited JSON over Unix domain socket.
 * Mirrors Go internal/cli/ipc.go exactly.
 *
 * @module
 */

import { createServer, type Server, type Socket } from 'node:net';
import { connect } from 'node:net';
import { unlink } from 'node:fs/promises';
import { chmodSync } from 'node:fs';
import { v4 as uuidv4 } from 'uuid';

// ---------------------------------------------------------------------------
// JSON-RPC 2.0 protocol types
// ---------------------------------------------------------------------------

/** JSON-RPC 2.0 request envelope. */
export interface IPCRequest {
  jsonrpc: '2.0';
  id: string;
  method: string;
  params?: unknown;
}

/** JSON-RPC 2.0 response envelope. */
export interface IPCResponse {
  jsonrpc: '2.0';
  id: string;
  result?: unknown;
  error?: IPCError;
}

/** JSON-RPC 2.0 error object. */
export interface IPCError {
  code: number;
  message: string;
  data?: unknown;
}

/** Handler function signature for IPC methods. */
export type IPCHandlerFunc = (req: IPCRequest) => Promise<IPCResponse>;

// ---------------------------------------------------------------------------
// JSON-RPC 2.0 error codes
// ---------------------------------------------------------------------------

/** Parse error: invalid JSON was received. */
export const ERR_PARSE = -32700;
/** Invalid Request: the JSON sent is not a valid Request object. */
export const ERR_INVALID_REQUEST = -32600;
/** Method not found. */
export const ERR_METHOD_NOT_FOUND = -32601;
/** Internal error (server-defined). */
export const ERR_SERVER = -32000;
/** Invalid params. */
export const ERR_INVALID_PARAMS = -32602;

// ---------------------------------------------------------------------------
// Factory helpers
// ---------------------------------------------------------------------------

/** Create a new IPCRequest with an auto-generated UUID v4 id. */
export function newIPCRequest(method: string, params?: unknown): IPCRequest {
  return {
    jsonrpc: '2.0',
    id: uuidv4(),
    method,
    ...(params !== undefined ? { params } : {}),
  };
}

/** Create a successful IPCResponse. */
export function newIPCResponse(id: string, result: unknown): IPCResponse {
  return { jsonrpc: '2.0', id, result };
}

/** Create an error IPCResponse. */
export function newIPCErrorResponse(id: string, code: number, message: string): IPCResponse {
  return {
    jsonrpc: '2.0',
    id,
    error: { code, message },
  };
}

// ---------------------------------------------------------------------------
// IPCServer
// ---------------------------------------------------------------------------

/**
 * IPCServer is a Unix domain socket server that dispatches JSON-RPC 2.0
 * requests to registered method handlers.
 *
 * Mirrors Go IPCServer in internal/cli/ipc.go.
 */
export class IPCServer {
  private readonly sockPath: string;
  private readonly handlers = new Map<string, IPCHandlerFunc>();
  private server: Server | null = null;
  private connections = new Set<Socket>();

  constructor(sockPath: string) {
    this.sockPath = sockPath;
  }

  /** Bind a handler to a method name. Must be called before start(). */
  register(method: string, handler: IPCHandlerFunc): void {
    this.handlers.set(method, handler);
  }

  /** Start accepting connections. Removes stale socket, chmod 0o600. */
  async start(): Promise<void> {
    // Remove stale socket.
    try {
      await unlink(this.sockPath);
    } catch {
      // May not exist — ignore.
    }

    return new Promise<void>((resolve, reject) => {
      this.server = createServer((conn) => {
        this.connections.add(conn);
        conn.on('close', () => this.connections.delete(conn));
        this.handleConn(conn);
      });

      this.server.on('error', (err) => {
        reject(new Error(`ipc server listen: ${err.message}`));
      });

      this.server.listen(this.sockPath, () => {
        try {
          chmodSync(this.sockPath, 0o600);
        } catch (err) {
          this.server?.close();
          reject(new Error(`ipc server chmod socket: ${(err as Error).message}`));
          return;
        }
        resolve();
      });
    });
  }

  /** Handle a single connection: read newline-delimited JSON, dispatch, respond. */
  private handleConn(conn: Socket): void {
    let buffer = '';

    conn.on('data', (chunk) => {
      buffer += chunk.toString('utf8');
      let newlineIdx: number;
      while ((newlineIdx = buffer.indexOf('\n')) !== -1) {
        const line = buffer.slice(0, newlineIdx).trim();
        buffer = buffer.slice(newlineIdx + 1);
        if (line.length === 0) continue;

        void this.dispatch(line).then((resp) => {
          const respJson = JSON.stringify(resp) + '\n';
          conn.write(respJson);
        });
      }
    });

    conn.on('error', () => {
      // Connection errors are transient — just close.
      conn.destroy();
    });
  }

  /** Parse a single JSON-RPC request line and invoke the handler. */
  private async dispatch(line: string): Promise<IPCResponse> {
    let req: IPCRequest;
    try {
      req = JSON.parse(line) as IPCRequest;
    } catch {
      return newIPCErrorResponse('', ERR_PARSE, 'Parse error');
    }

    if (req.jsonrpc !== '2.0') {
      return newIPCErrorResponse(req.id ?? '', ERR_INVALID_REQUEST, 'Invalid Request');
    }

    const handler = this.handlers.get(req.method);
    if (!handler) {
      return newIPCErrorResponse(req.id, ERR_METHOD_NOT_FOUND, 'Method not found');
    }

    // Handler must catch its own errors and return error responses.
    try {
      return await handler(req);
    } catch (err) {
      return newIPCErrorResponse(req.id, ERR_SERVER, (err as Error).message);
    }
  }

  /** Stop the server and close all connections. */
  async stop(): Promise<void> {
    // Close all active connections.
    for (const conn of this.connections) {
      conn.destroy();
    }
    this.connections.clear();

    return new Promise<void>((resolve) => {
      if (!this.server) {
        resolve();
        return;
      }
      this.server.close(() => resolve());
    });
  }
}

// ---------------------------------------------------------------------------
// IPCClient
// ---------------------------------------------------------------------------

/**
 * IPCClient connects to a running daemon over a Unix domain socket.
 * Each call() opens a fresh connection, sends one request, reads one
 * response, and closes.
 *
 * Mirrors Go IPCClient in internal/cli/ipc.go.
 */
export class IPCClient {
  private readonly sockPath: string;
  private readonly timeout: number;

  constructor(sockPath: string, timeout = 5000) {
    this.sockPath = sockPath;
    this.timeout = timeout;
  }

  /**
   * Send a single JSON-RPC request and return the response.
   * Opens a new connection per call.
   */
  async call(method: string, params?: unknown): Promise<IPCResponse> {
    const req = newIPCRequest(method, params);
    const reqJson = JSON.stringify(req) + '\n';

    return new Promise<IPCResponse>((resolve, reject) => {
      const conn = connect({ path: this.sockPath });
      let responseBuffer = '';
      let settled = false;

      const timer = setTimeout(() => {
        if (!settled) {
          settled = true;
          conn.destroy();
          reject(new Error('ipc client: connection timed out'));
        }
      }, this.timeout);

      conn.on('connect', () => {
        conn.write(reqJson);
      });

      conn.on('data', (chunk) => {
        responseBuffer += chunk.toString('utf8');
        const newlineIdx = responseBuffer.indexOf('\n');
        if (newlineIdx !== -1) {
          const line = responseBuffer.slice(0, newlineIdx).trim();
          clearTimeout(timer);
          settled = true;
          conn.end();
          try {
            const resp = JSON.parse(line) as IPCResponse;
            resolve(resp);
          } catch (err) {
            reject(new Error(`ipc client unmarshal response: ${(err as Error).message}`));
          }
        }
      });

      conn.on('error', (err) => {
        if (!settled) {
          clearTimeout(timer);
          settled = true;
          reject(new Error(`ipc client dial: ${err.message}`));
        }
      });

      conn.on('close', () => {
        clearTimeout(timer);
        if (!settled) {
          settled = true;
          reject(new Error('ipc client: connection closed before response'));
        }
      });
    });
  }
}
